openhcl/virt_mshv_vtl/src/processor/mod.rs

```diff
@@ -678,8 +684,29 @@ impl<'p, T: Backing> Processor for UhProcessor<'p, T> {
                     }
                 }
 
-                // TODO WHP GUEST VSM: This should be next_vtl
-                if T::halt_in_usermode(self, GuestVtl::Vtl0) {
+                match self.exit_vtl {
+                    GuestVtl::Vtl0 => {
+                        // Check for VTL preemption
+                        if let Some(vector) = interrupt_pending[GuestVtl::Vtl1] {
+                            let priority = vector >> 4;
+                            // TODO actually check priority registers
+                            if priority > 0 {
+                                T::switch_vtl_state(self, GuestVtl::Vtl0, GuestVtl::Vtl1);
```

Comment:
This could result in us switching state twice, if we handle a VTL 1 return to VTL 0, and then have to immediately preempt that with an interrupt. But that's just a perf issue. We could do some extra wiring to delay state switching until the absolute last moment if we have to.